File: 07_QUALITY_ASSURANCE/test_candidate_ingest_fail_closed_contract.py
#!/usr/bin/env python3
"""
JAYT CORP — CANDIDATE INGESTION FAIL-CLOSED CONTRACT TEST SUITE
WORK ORDER: JAYT-INGESTION-SAFETY-029A (TRUE 2-FILE ATOMIC RECOVERY)
"""

import hashlib
import json
import os
import shutil
import subprocess
import sys
import tempfile
import unittest

if sys.platform == 'win32':
    try:
        sys.stdout.reconfigure(encoding='utf-8')
        sys.stderr.reconfigure(encoding='utf-8')
    except Exception:
        pass

QA_DIR = os.path.dirname(os.path.abspath(__file__))
REPO_ROOT = os.path.abspath(os.path.join(QA_DIR, '..'))
DEALS_DIR = os.path.join(REPO_ROOT, '05_DEAL_AND_AFFILIATE')
sys.path.insert(0, DEALS_DIR)

from ingest_candidate_to_catalog import ingest_candidate_file
from catalog_workflow_validator import compute_file_sha256

LIVE_FEED_PATH = os.path.join(DEALS_DIR, 'deals_feed.json')
METIZ_CANDIDATE_PATH = os.path.join(DEALS_DIR, 'candidates', 'pending_review', 'candidate_18_CAND-DNG-METIZ-U22-REAL.json')
INGEST_SCRIPT_PATH = os.path.join(DEALS_DIR, 'ingest_candidate_to_catalog.py')


class TestCandidateIngestFailClosedContract(unittest.TestCase):
    """Test suite proving that unconfirmed temporal validity candidates are strictly rejected by the real ingestion pipeline."""

    @classmethod
    def setUpClass(cls):
        cls.live_pre_hash = compute_file_sha256(LIVE_FEED_PATH)

    def setUp(self):
        self.temp_dir = tempfile.mkdtemp(prefix='jayt_py_ingest_test_')
        self.isolated_feed = os.path.join(self.temp_dir, 'test_deals_feed.json')
        self.isolated_evidence = os.path.join(self.temp_dir, 'test_evidence_store.json')
        with open(self.isolated_feed, 'w', encoding='utf-8') as f:
            json.dump([], f, indent=2)
        with open(self.isolated_evidence, 'w', encoding='utf-8') as f:
            json.dump({}, f, indent=2)
        self.pre_feed_hash = compute_file_sha256(self.isolated_feed)
        self.pre_evidence_hash = compute_file_sha256(self.isolated_evidence)

    def tearDown(self):
        shutil.rmtree(self.temp_dir, ignore_errors=True)

    def test_01_real_metiz_unconfirmed_blocked_from_ingestion(self):
        """Proof: Real Metiz candidate with UNCONFIRMED_AT_CAPTURE_TIME fails ingestion and preserves 0 items."""
        res = ingest_candidate_file(METIZ_CANDIDATE_PATH, self.isolated_feed, self.isolated_evidence)
        self.assertFalse(res['success'], "Ingestion must return success=False")
        self.assertIn("REJECTED_UNCONFIRMED_TEMPORAL_VALIDITY", res['message'])

        post_feed_hash = compute_file_sha256(self.isolated_feed)
        post_evidence_hash = compute_file_sha256(self.isolated_evidence)
        self.assertEqual(self.pre_feed_hash, post_feed_hash, "Target deals feed must have identical SHA-256 byte-for-byte")
        self.assertEqual(self.pre_evidence_hash, post_evidence_hash, "Target evidence store must have identical SHA-256 byte-for-byte")

        with open(self.isolated_feed, 'r', encoding='utf-8') as f:
            feed_data = json.load(f)
        self.assertEqual(len(feed_data), 0, "Feed must contain 0 deals")

    def test_02_synthetic_fixture_unconfirmed_blocked(self):
        """Proof: Synthetic candidate with UNCONFIRMED_AT_CAPTURE_TIME is blocked from ingestion."""
        fixture_path = os.path.join(self.temp_dir, 'synthetic_candidate.json')
        synthetic_data = {
            "evidence": {
                "EVID_SYNTHETIC": {
                    "deal_id": "DNG-SYNTHETIC-01",
                    "source_url": "https://metiz.vn/promo",
                    "source_type": "OFFICIAL_PROMOTION_ANNOUNCEMENT",
                    "recorded_by": "JAYT_DESK_REVIEW_PUBLIC_SOURCE",
                    "verification_status": "NEEDS_RECHECK",
                    "temporal_validity": "UNCONFIRMED_AT_CAPTURE_TIME",
                    "expiry_basis": "NOT_OBSERVED_ON_CAPTURED_PROMOTION_PAGE",
                    "checked_at": "2026-08-22T00:00:00Z",
                    "captured_at": "2026-08-22T00:00:00Z",
                    "notes": "Test synthetic"
                }
            },
            "deals": [
                {
                    "deal_id": "DNG-SYNTHETIC-01",
                    "title": "Deal Test",
                    "merchant": "Metiz",
                    "zone": "ZONE_HELIO_METIZ",
                    "category": "local_entertainment",
                    "need_collection": "general",
                    "budget_tier": "under_100k",
                    "group_size": "1_person",
                    "contextual_reason": "Test",
                    "taxonomy": "PROBING",
                    "affiliate_type": "DIRECT_DEAL",
                    "source_url": "https://metiz.vn/promo",
                    "evidence_ref": "EVID_SYNTHETIC",
                    "disclosure": "Test",
                    "lifecycle_status": "PROBING",
                    "category_scope": "LOCAL_EXPERIENCE"
                }
            ]
        }
        with open(fixture_path, 'w', encoding='utf-8') as f:
            json.dump(synthetic_data, f, indent=2)

        res = ingest_candidate_file(fixture_path, self.isolated_feed, self.isolated_evidence)
        self.assertFalse(res['success'])
        self.assertEqual(res['error_code'], 'ERR_UNCONFIRMED_TEMPORAL_VALIDITY')

        post_feed_hash = compute_file_sha256(self.isolated_feed)
        post_evidence_hash = compute_file_sha256(self.isolated_evidence)
        self.assertEqual(self.pre_feed_hash, post_feed_hash)
        self.assertEqual(self.pre_evidence_hash, post_evidence_hash)

    def test_03_pass_deal_without_ceo_approval_blocked(self):
        """Proof: Ready/Pass candidate without explicit CEO approval manifest is strictly BLOCKED from ingestion."""
        fixture_path = os.path.join(self.temp_dir, 'ready_deal.json')
        ready_data = {
            "evidence": {
                "EVID_READY": {
                    "deal_id": "DNG-READY-01",
                    "source_url": "https://metiz.vn/promotion/khuyen-mai-gia-ve-u22-21.html",
                    "source_type": "OFFICIAL_PROMOTION_ANNOUNCEMENT",
                    "recorded_by": "JAYT_DESK_REVIEW_PUBLIC_SOURCE",
                    "verification_status": "READY_FOR_CEO_REVIEW",
                    "checked_at": "2026-08-22T00:00:00Z",
                    "captured_at": "2026-08-22T00:00:00Z",
                    "notes": "Valid test"
                }
            },
            "deals": [
                {
                    "deal_id": "DNG-READY-01",
                    "title": "Deal Ready",
                    "merchant": "Metiz Cinema",
                    "zone": "ZONE_HELIO_METIZ",
                    "category": "local_entertainment",
                    "need_collection": "general",
                    "budget_tier": "under_100k",
                    "group_size": "1_person",
                    "contextual_reason": "Test",
                    "taxonomy": "PROBING",
                    "affiliate_type": "DIRECT_DEAL",
                    "source_url": "https://metiz.vn/promotion/khuyen-mai-gia-ve-u22-21.html",
                    "evidence_ref": "EVID_READY",
                    "disclosure": "Test",
                    "lifecycle_status": "PROBING",
                    "category_scope": "LOCAL_EXPERIENCE"
                }
            ]
        }
        with open(fixture_path, 'w', encoding='utf-8') as f:
            json.dump(ready_data, f, indent=2)

        res = ingest_candidate_file(fixture_path, self.isolated_feed, self.isolated_evidence)
        self.assertFalse(res['success'])
        self.assertEqual(res['error_code'], 'ERR_CEO_APPROVAL_REQUIRED')

        post_feed_hash = compute_file_sha256(self.isolated_feed)
        post_evidence_hash = compute_file_sha256(self.isolated_evidence)
        self.assertEqual(self.pre_feed_hash, post_feed_hash)
        self.assertEqual(self.pre_evidence_hash, post_evidence_hash)

    def test_04a_fault_injection_phase1_crash_rollback(self):
        """Proof: Crash during Phase 1 evidence write triggers atomic rollback; both feeds retain exact pre-test SHA-256."""
        fixture_path = os.path.join(self.temp_dir, 'ready_deal.json')
        ready_data = {
            "evidence": {
                "EVID_READY": {
                    "deal_id": "DNG-READY-01",
                    "source_url": "https://metiz.vn/promotion/khuyen-mai-gia-ve-u22-21.html",
                    "source_type": "OFFICIAL_PROMOTION_ANNOUNCEMENT",
                    "recorded_by": "JAYT_DESK_REVIEW_PUBLIC_SOURCE",
                    "verification_status": "READY_FOR_CEO_REVIEW",
                    "checked_at": "2026-08-22T00:00:00Z",
                    "captured_at": "2026-08-22T00:00:00Z",
                    "notes": "Valid test"
                }
            },
            "deals": [
                {
                    "deal_id": "DNG-READY-01",
                    "title": "Deal Ready",
                    "merchant": "Metiz Cinema",
                    "zone": "ZONE_HELIO_METIZ",
                    "category": "local_entertainment",
                    "need_collection": "general",
                    "budget_tier": "under_100k",
                    "group_size": "1_person",
                    "contextual_reason": "Test",
                    "taxonomy": "PROBING",
                    "affiliate_type": "DIRECT_DEAL",
                    "source_url": "https://metiz.vn/promotion/khuyen-mai-gia-ve-u22-21.html",
                    "evidence_ref": "EVID_READY",
                    "disclosure": "Test",
                    "lifecycle_status": "PROBING",
                    "category_scope": "LOCAL_EXPERIENCE"
                }
            ]
        }
        with open(fixture_path, 'w', encoding='utf-8') as f:
            json.dump(ready_data, f, indent=2)

        mock_approval = {
            "work_order": "JAYT-TEST-001",
            "candidate_id": "DNG-READY-01",
            "approved_by": "CEO_JAY_TRAN",
            "status": "AUTHORIZED_FOR_CATALOG_INGESTION"
        }

        res = ingest_candidate_file(
            fixture_path,
            self.isolated_feed,
            self.isolated_evidence,
            approval_manifest=mock_approval,
            simulate_evidence_write_failure=True
        )

        self.assertFalse(res['success'])
        self.assertEqual(res['error_code'], 'ERR_INGESTION_ROLLBACK')

        post_feed_hash = compute_file_sha256(self.isolated_feed)
        post_evidence_hash = compute_file_sha256(self.isolated_evidence)
        self.assertEqual(self.pre_feed_hash, post_feed_hash, "Feed must be completely rolled back")
        self.assertEqual(self.pre_evidence_hash, post_evidence_hash, "Evidence store must be completely rolled back")

    def test_04b_fault_injection_post_feed_swap_recovery_proof(self):
        """Proof: Crash AFTER feed swap restores feed from backup copy; 0 partial mutation."""
        fixture_path = os.path.join(self.temp_dir, 'ready_deal.json')
        ready_data = {
            "evidence": {
                "EVID_READY": {
                    "deal_id": "DNG-READY-01",
                    "source_url": "https://metiz.vn/promotion/khuyen-mai-gia-ve-u22-21.html",
                    "source_type": "OFFICIAL_PROMOTION_ANNOUNCEMENT",
                    "recorded_by": "JAYT_DESK_REVIEW_PUBLIC_SOURCE",
                    "verification_status": "READY_FOR_CEO_REVIEW",
                    "checked_at": "2026-08-22T00:00:00Z",
                    "captured_at": "2026-08-22T00:00:00Z",
                    "notes": "Valid test"
                }
            },
            "deals": [
                {
                    "deal_id": "DNG-READY-01",
                    "title": "Deal Ready",
                    "merchant": "Metiz Cinema",
                    "zone": "ZONE_HELIO_METIZ",
                    "category": "local_entertainment",
                    "need_collection": "general",
                    "budget_tier": "under_100k",
                    "group_size": "1_person",
                    "contextual_reason": "Test",
                    "taxonomy": "PROBING",
                    "affiliate_type": "DIRECT_DEAL",
                    "source_url": "https://metiz.vn/promotion/khuyen-mai-gia-ve-u22-21.html",
                    "evidence_ref": "EVID_READY",
                    "disclosure": "Test",
                    "lifecycle_status": "PROBING",
                    "category_scope": "LOCAL_EXPERIENCE"
                }
            ]
        }
        with open(fixture_path, 'w', encoding='utf-8') as f:
            json.dump(ready_data, f, indent=2)

        mock_approval = {
            "work_order": "JAYT-TEST-001",
            "candidate_id": "DNG-READY-01",
            "approved_by": "CEO_JAY_TRAN",
            "status": "AUTHORIZED_FOR_CATALOG_INGESTION"
        }

        res = ingest_candidate_file(
            fixture_path,
            self.isolated_feed,
            self.isolated_evidence,
            approval_manifest=mock_approval,
            simulate_post_feed_swap_failure=True
        )

        self.assertFalse(res['success'])
        self.assertEqual(res['error_code'], 'ERR_INGESTION_ROLLBACK')

        post_feed_hash = compute_file_sha256(self.isolated_feed)
        post_evidence_hash = compute_file_sha256(self.isolated_evidence)
        self.assertEqual(self.pre_feed_hash, post_feed_hash, "Feed must be restored from backup file")
        self.assertEqual(self.pre_evidence_hash, post_evidence_hash, "Evidence store must remain intact")

        with open(self.isolated_feed, 'r', encoding='utf-8') as f:
            feed_data = json.load(f)
        self.assertEqual(len(feed_data), 0, "Feed must contain 0 items after rollback")

    def test_05_cli_subprocess_exits_non_zero(self):
        """Proof: CLI subprocess for ingestion pipeline exits with exit code 1 on unconfirmed temporal candidate."""
        python_exe = sys.executable
        res = subprocess.run([
            python_exe,
            INGEST_SCRIPT_PATH,
            '--candidate', METIZ_CANDIDATE_PATH,
            '--target-feed', self.isolated_feed,
            '--target-evidence', self.isolated_evidence
        ], capture_output=True, text=True, encoding='utf-8')

        self.assertEqual(res.returncode, 1, f"Expected non-zero exit code 1, got {res.returncode}")
        self.assertIn("REJECTED_UNCONFIRMED_TEMPORAL_VALIDITY", res.stdout)

        post_feed_hash = compute_file_sha256(self.isolated_feed)
        post_evidence_hash = compute_file_sha256(self.isolated_evidence)
        self.assertEqual(self.pre_feed_hash, post_feed_hash)
        self.assertEqual(self.pre_evidence_hash, post_evidence_hash)

    def test_06_live_catalog_unmutated_invariant(self):
        """Proof: Live deals_feed.json remains completely unmutated [] (4f53cda18c2baa0c0354bb5f9a3ecbe5ed12ab4d8e11ba873c2f11161202b945)."""
        live_post_hash = compute_file_sha256(LIVE_FEED_PATH)
        self.assertEqual(self.live_pre_hash, live_post_hash)
        self.assertEqual(live_post_hash, '4f53cda18c2baa0c0354bb5f9a3ecbe5ed12ab4d8e11ba873c2f11161202b945')

        with open(LIVE_FEED_PATH, 'r', encoding='utf-8') as f:
            live_data = json.load(f)
        self.assertEqual(len(live_data), 0)


if __name__ == '__main__':
    suite = unittest.TestLoader().loadTestsFromTestCase(TestCandidateIngestFailClosedContract)
    runner = unittest.TextTestRunner(verbosity=2)
    result = runner.run(suite)
    sys.exit(0 if result.wasSuccessful() else 1)

File: 07_QUALITY_ASSURANCE/test_candidate_ingest_fail_closed_contract.js
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const os = require('os');
const { spawnSync } = require('child_process');

const repoRoot = path.resolve(__dirname, '..');
const ingestPipelineJs = path.join(repoRoot, '05_DEAL_AND_AFFILIATE', 'ingest_candidate_to_catalog.js');
const { ingestCandidateFile } = require(ingestPipelineJs);

const liveFeedPath = path.join(repoRoot, '05_DEAL_AND_AFFILIATE', 'deals_feed.json');
const metizCandidatePath = path.join(repoRoot, '05_DEAL_AND_AFFILIATE', 'candidates', 'pending_review', 'candidate_18_CAND-DNG-METIZ-U22-REAL.json');

console.log('🧪 [JAYT-INGEST-SAFETY-TEST] Khởi chạy bộ kiểm thử An Toàn Ingestion Pipeline & True 2-File Atomic 2PC (029A)...');

let allPassed = true;
function assertTest(name, condition, detail = '') {
  console.log(`  [${name}]: [${condition ? 'PASS' : 'FAIL'}]${detail ? ' - ' + detail : ''}`);
  if (!condition) allPassed = false;
}

function getSha256(filePath) {
  if (!fs.existsSync(filePath)) return null;
  return crypto.createHash('sha256').update(fs.readFileSync(filePath)).digest('hex');
}

const liveFeedPreHash = getSha256(liveFeedPath);
const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'jayt_ingest_safety_test_'));

try {
  const isolatedFeedPath = path.join(tempDir, 'test_deals_feed.json');
  const isolatedEvidencePath = path.join(tempDir, 'test_evidence_store.json');
  fs.writeFileSync(isolatedFeedPath, JSON.stringify([], null, 2), 'utf8');
  fs.writeFileSync(isolatedEvidencePath, JSON.stringify({}, null, 2), 'utf8');

  const preFeedHash = getSha256(isolatedFeedPath);
  const preEvidenceHash = getSha256(isolatedEvidencePath);

  // ---------------------------------------------------------------------------
  // [INGEST_01] Real Candidate Metiz with UNCONFIRMED_AT_CAPTURE_TIME is Blocked
  // ---------------------------------------------------------------------------
  const res1 = ingestCandidateFile(metizCandidatePath, isolatedFeedPath, isolatedEvidencePath);
  const postFeedHash1 = getSha256(isolatedFeedPath);
  const postEvidenceHash1 = getSha256(isolatedEvidencePath);
  const feedContent1 = JSON.parse(fs.readFileSync(isolatedFeedPath, 'utf8'));

  assertTest(
    'INGEST_01_REAL_METIZ_UNCONFIRMED_BLOCKED',
    res1.success === false &&
    res1.error_code === 'ERR_UNCONFIRMED_TEMPORAL_VALIDITY' &&
    preFeedHash === postFeedHash1 &&
    preEvidenceHash === postEvidenceHash1 &&
    feedContent1.length === 0,
    `Refusal: ${res1.message}`
  );

  // ---------------------------------------------------------------------------
  // [INGEST_02] Synthetic Fixture with UNCONFIRMED_AT_CAPTURE_TIME is Blocked
  // ---------------------------------------------------------------------------
  const fixtureCandidatePath = path.join(tempDir, 'candidate_synthetic_unconfirmed.json');
  const syntheticData = {
    evidence: {
      "EVID_SYNTHETIC": {
        deal_id: "DNG-SYNTHETIC-01",
        source_url: "https://metiz.vn/promo",
        source_type: "OFFICIAL_PROMOTION_ANNOUNCEMENT",
        recorded_by: "JAYT_DESK_REVIEW_PUBLIC_SOURCE",
        verification_status: "NEEDS_RECHECK",
        temporal_validity: "UNCONFIRMED_AT_CAPTURE_TIME",
        expiry_basis: "NOT_OBSERVED_ON_CAPTURED_PROMOTION_PAGE",
        checked_at: new Date().toISOString(),
        captured_at: new Date().toISOString(),
        notes: "Test synthetic"
      }
    },
    deals: [
      {
        deal_id: "DNG-SYNTHETIC-01",
        title: "Deal Chưa Rõ Thời Hạn",
        merchant: "Metiz Cinema",
        zone: "ZONE_HELIO_METIZ",
        category: "local_entertainment",
        need_collection: "general",
        budget_tier: "under_100k",
        group_size: "1_person",
        contextual_reason: "Test",
        taxonomy: "PROBING",
        affiliate_type: "DIRECT_DEAL",
        source_url: "https://metiz.vn/promo",
        evidence_ref: "EVID_SYNTHETIC",
        disclosure: "Test",
        lifecycle_status: "PROBING",
        category_scope: "LOCAL_EXPERIENCE"
      }
    ]
  };
  fs.writeFileSync(fixtureCandidatePath, JSON.stringify(syntheticData, null, 2), 'utf8');

  const res2 = ingestCandidateFile(fixtureCandidatePath, isolatedFeedPath, isolatedEvidencePath);
  const postFeedHash2 = getSha256(isolatedFeedPath);
  const feedContent2 = JSON.parse(fs.readFileSync(isolatedFeedPath, 'utf8'));

  assertTest(
    'INGEST_02_FIXTURE_WITH_UNCONFIRMED_TEMPORAL_BLOCKED',
    res2.success === false &&
    res2.error_code === 'ERR_UNCONFIRMED_TEMPORAL_VALIDITY' &&
    preFeedHash === postFeedHash2 &&
    feedContent2.length === 0,
    `Refusal: ${res2.message}`
  );

  // ---------------------------------------------------------------------------
  // [INGEST_03] Positive Deal Attempt without CEO Approval Manifest is Blocked
  // ---------------------------------------------------------------------------
  const candidateReadyForReview = {
    candidate_id: "CAND-TEST-READY-03",
    merchant: "Metiz Cinema Đà Nẵng",
    title: "Vé 2D 55K U22",
    category: "local_entertainment",
    category_scope: "LOCAL_EXPERIENCE",
    affiliate_type: "DIRECT_DEAL",
    source_url: "https://metiz.vn/promotion/khuyen-mai-gia-ve-u22-21.html",
    artifact_source_url: "https://metiz.vn/promotion/khuyen-mai-gia-ve-u22-21.html",
    captured_at: new Date().toISOString(),
    artifact_screenshot: "metiz_u22_official_promo_capture.png",
    artifact_screenshot_hash: "5c563aebfe78b487fdcf2f1c1f13e7c07ea447a52aff5873320d549b6d034ea7",
    artifact_text_dump: "metiz_u22_official_promo_text.txt",
    artifact_text_hash: "c17418d5d44c7c92a154dd894a06249f3d5fbd7cab8b30366375abdcf73b59b0",
    extracted_claims: {
      price_snippet: "55.000đ",
      conditions_snippet: "thành viên Metiz Cinema từ 22 tuổi trở xuống",
      expiry_snippet: "thứ Ba đến thứ Năm"
    },
    observed_price_or_offer: "55.000đ",
    observed_conditions: "thành viên Metiz Cinema từ 22 tuổi trở xuống",
    expiry_basis: "Áp dụng Thứ Ba đến Thứ Năm",
    verification_readiness: "READY_FOR_CEO_REVIEW"
  };
  const candidateReadyFile = path.join(tempDir, 'candidate_ready_no_approval.json');
  fs.writeFileSync(candidateReadyFile, JSON.stringify(candidateReadyForReview, null, 2), 'utf8');

  const res3 = ingestCandidateFile(candidateReadyFile, isolatedFeedPath, isolatedEvidencePath);
  const postFeedHash3 = getSha256(isolatedFeedPath);
  assertTest(
    'INGEST_03_PASS_WITHOUT_CEO_APPROVAL_BLOCKED',
    res3.success === false &&
    res3.error_code === 'ERR_CEO_APPROVAL_REQUIRED' &&
    preFeedHash === postFeedHash3,
    `Refusal: ${res3.message}`
  );

  // ---------------------------------------------------------------------------
  // [INGEST_04a] Fault Injection Phase 1 Crash Rollback Proof
  // ---------------------------------------------------------------------------
  const mockApproval = {
    work_order: "JAYT-TEST-001",
    candidate_id: "CAND-TEST-READY-03",
    approved_by: "CEO_JAY_TRAN",
    status: "AUTHORIZED_FOR_CATALOG_INGESTION"
  };

  const res4a = ingestCandidateFile(candidateReadyFile, isolatedFeedPath, isolatedEvidencePath, {
    approvalManifest: mockApproval,
    simulate_evidence_write_failure: true
  });

  const postFeedHash4a = getSha256(isolatedFeedPath);
  const postEvidenceHash4a = getSha256(isolatedEvidencePath);

  assertTest(
    'INGEST_04a_FAULT_INJECTION_PHASE1_CRASH_ROLLED_BACK',
    res4a.success === false &&
    res4a.error_code === 'ERR_INGESTION_ROLLBACK' &&
    preFeedHash === postFeedHash4a &&
    preEvidenceHash === postEvidenceHash4a,
    `Phase 1 rollback verified: Feed and evidence store byte-for-byte identical to pre-test state.`
  );

  // ---------------------------------------------------------------------------
  // [INGEST_04b] True 2-File Atomic Recovery: Crash AFTER Feed Swap Rollback Proof
  // ---------------------------------------------------------------------------
  const res4b = ingestCandidateFile(candidateReadyFile, isolatedFeedPath, isolatedEvidencePath, {
    approvalManifest: mockApproval,
    simulate_post_feed_swap_failure: true
  });

  const postFeedHash4b = getSha256(isolatedFeedPath);
  const postEvidenceHash4b = getSha256(isolatedEvidencePath);
  const feedContent4b = JSON.parse(fs.readFileSync(isolatedFeedPath, 'utf8'));

  assertTest(
    'INGEST_04b_POST_FEED_SWAP_FAILURE_RESTORED_FROM_BACKUP',
    res4b.success === false &&
    res4b.error_code === 'ERR_INGESTION_ROLLBACK' &&
    preFeedHash === postFeedHash4b &&
    preEvidenceHash === postEvidenceHash4b &&
    feedContent4b.length === 0,
    `True 2-File Atomic Recovery verified: Feed was restored from backup after simulated crash before evidence commit.`
  );

  // ---------------------------------------------------------------------------
  // [INGEST_05] Real Ingest CLI Subprocess Exits with Non-Zero Exit Code 1
  // ---------------------------------------------------------------------------
  const proc = spawnSync('node', [
    ingestPipelineJs,
    '--candidate', metizCandidatePath,
    '--target-feed', isolatedFeedPath,
    '--target-evidence', isolatedEvidencePath
  ], { encoding: 'utf8' });

  const postFeedHash5 = getSha256(isolatedFeedPath);
  const feedContent5 = JSON.parse(fs.readFileSync(isolatedFeedPath, 'utf8'));

  assertTest(
    'INGEST_05_CLI_SUBPROCESS_NON_ZERO_EXIT',
    proc.status === 1 &&
    proc.stdout.includes('REJECTED_UNCONFIRMED_TEMPORAL_VALIDITY') &&
    preFeedHash === postFeedHash5 &&
    feedContent5.length === 0,
    `Exit Code: ${proc.status}, Output contains refusal message`
  );

  // ---------------------------------------------------------------------------
  // [INGEST_06] Live Catalog Deals Feed Zero-Mutation Invariant Proof
  // ---------------------------------------------------------------------------
  const liveFeedPostHash = getSha256(liveFeedPath);
  const liveDeals = JSON.parse(fs.readFileSync(liveFeedPath, 'utf8'));

  assertTest(
    'INGEST_06_LIVE_CATALOG_ZERO_MUTATION_PROOF',
    liveFeedPreHash === liveFeedPostHash &&
    liveDeals.length === 0 &&
    liveFeedPostHash === '4f53cda18c2baa0c0354bb5f9a3ecbe5ed12ab4d8e11ba873c2f11161202b945',
    `Live deals_feed.json remained untouched []. SHA-256: ${liveFeedPostHash}`
  );

} finally {
  try { fs.rmSync(tempDir, { recursive: true, force: true }); } catch {}
}

console.log('\n' + (allPassed ? '🟢' : '❌') + ' [INGEST-SAFETY-SUMMARY] Toàn bộ ' + (allPassed ? '7/7' : 'bài') + ' kiểm thử An Toàn Ingestion Pipeline & True 2-File Atomic Recovery đã ' + (allPassed ? 'ĐẠT [PASS]' : 'THẤT BẠI [FAIL]') + '!');

if (!allPassed) {
  process.exit(1);
} else {
  process.exit(0);
}
